contracts/strategies/RFPStrategy.sol

```diff
@@ -0,0 +1,278 @@
+
+    /// @notice Update max bid for RFP pool
+    /// @param _maxBid The max bid to be set
+    function updateMaxBid(uint256 _maxBid) external onlyPoolManager(msg.sender) {
+        maxBid = _maxBid;
+
+        emit MAX_BID_UPDATED(maxBid);
+    }
+
+    /// @notice Reject pending milestone
+    /// @param _milestoneId Id of the milestone
+    function rejectMilestone(uint256 _milestoneId) external onlyPoolManager(msg.sender) {
+        if (milestones[_milestoneId].milestoneStatus == RecipientStatus.Accepted) {
+            revert MILESTONE_ALREADY_ACCEPTED();
+        }
+
+        milestones[_milestoneId].milestoneStatus = RecipientStatus.Rejected;
+        emit MILESTONE_REJECTED(_milestoneId);
+    }
+
+    /// ====================================
+    /// ============ Internal ==============
+    /// ====================================
+
+    /// @notice Submit proposal to RFP pool
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the transaction
+    function _registerRecipient(bytes memory _data, address _sender) internal override returns (address recipientId) {
+        if (acceptedRecipientId != address(0)) {
+            revert RECIPIENT_ALREADY_ACCEPTED();
+        }
+
+        address recipientAddress;
+        uint256 proposalBid;
+        Metadata memory metadata;
+
+        // decode data custom to this strategy
+        if (registryGating) {
+            (recipientId, recipientAddress, proposalBid, metadata) =
+                abi.decode(_data, (address, address, uint256, Metadata));
+
+            if (!_isIdentityManager(recipientId, _sender)) {
+                revert UNAUTHORIZED();
+            }
+        } else {
+            (recipientAddress, proposalBid, metadata) = abi.decode(_data, (address, uint256, Metadata));
+            recipientId = _sender;
+        }
+
+        // check if proposal bid is less than max bid
+        if (proposalBid > maxBid) {
+            revert EXCEEDING_MAX_BID();
+        } else if (proposalBid == 0) {
+            proposalBid = maxBid;
+        }
+
+        Recipient memory recipient = Recipient({
+            recipientAddress: recipientAddress,
+            proposalBid: proposalBid,
+            recipientStatus: RecipientStatus.Pending
+        });
+
+        recipients[recipientId] = recipient;
+        _recipientIds.push(recipientId);
+
+        emit Registered(recipientId, _data, _sender);
+    }
+
+    /// @notice Select recipient for RFP allocation
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the allocation
+    function _allocate(bytes memory _data, address _sender) internal override nonReentrant onlyPoolManager(_sender) {
+        if (acceptedRecipientId != address(0)) {
+            revert RECIPIENT_ALREADY_ACCEPTED();
+        }
+
+        acceptedRecipientId = abi.decode(_data, (address));
+
+        uint256 _recipientsCounter = _recipientIds.length;
+        for (uint256 i = 0; i < _recipientsCounter;) {
```

Comment:
this for loop can end up in a DOS condition, if the `_recipientsCounter` reaches a certain limit

i would remove the entire for loop and change `getRecipientStatus` to
```solidity
function getRecipientStatus(address _recipientId) public view override returns (RecipientStatus status) {
    Recipient memory recipient = recipients[_recipientId];
    
    if (acceptedRecipientId == address(0) || acceptedRecipientId == _recipientId) {
        return recipient.recipientStatus;
    }
    
    return recipient.recipientStatus > RecipientStatus.None ? RecipientStatus.Rejected : RecipientStatus.None;
}
```

also i would make the `recipients`-mapping private and introduce a getter which replaces the status based on the logic above.

this enables us to remove the for loop entirely